### PR TITLE
Add tests for TypedArrays by using indexing API.

### PR DESCRIPTION
**Pull request by imiklos**
JerryScript-DCO-1.0-Signed-off-by: Istvan Miklos imiklos2@inf.u-szeged.hu

### CODE REVIEW

File: tests/unit-core/test-typedarray.c [hunk 1]
```diff
@@ -303,6 +303,44 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+
+  jerry_value_t ja = jerry_create_number (3);
+  jerry_value_t jb = jerry_create_number (6);
+  jerry_value_t jc = jerry_create_number (50);
+  jerry_value_t global_obj_val = jerry_get_global_object ();
+
+  for (uint32_t i = 0; test_entries[i].constructor_name != NULL; i++)
+  {
+    jerry_value_t prop_name = jerry_create_string ((const jerry_char_t *) test_entries[i].constructor_name);
+    jerry_value_t prop_value = jerry_get_property (global_obj_val, prop_name);
```
Comment (galpeter) [inline tests/unit-core/test-typedarray.c]: There is no need to access the typedarray via the global object, we have a typedarray constructor function: http://jerryscript.net/api-reference/#jerry_create_typedarray.

File: tests/unit-core/test-typedarray.c [hunk 2]
```diff
@@ -303,6 +303,44 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+    jerry_value_t length_arg = jerry_create_number (test_entries[i].element_count);
+
+    jerry_value_t typedarray = jerry_construct_object (prop_value, &length_arg, 1);
+
+    jerry_release_value (prop_name);
+    jerry_release_value (prop_value);
+    jerry_release_value (length_arg);
+
+    jerry_set_property_by_index (typedarray, 0, ja);
```
Comment (galpeter) [inline tests/unit-core/test-typedarray.c]: The `..set_property_...` call returns a value, which we also need to check.

File: tests/unit-core/test-typedarray.c [hunk 3]
```diff
@@ -303,6 +303,44 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+    jerry_set_property_by_index (typedarray, 1, jb);
+    jerry_set_property_by_index (typedarray, 2, jc);
+
+    TEST_ASSERT (jerry_get_number_value (ja) == jerry_get_number_value (jerry_get_property_by_index (typedarray, 0)));
```
Comment (galpeter) [inline tests/unit-core/test-typedarray.c]: As the `jerry_get_property_by_index` returns a `jerry_value_t` we need to release that (see: http://jerryscript.net/api-reference/#jerry_get_property_by_index)

File: tests/unit-core/test-typedarray.c [hunk 4]
```diff
@@ -303,6 +303,44 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+    TEST_ASSERT (jerry_get_number_value (jc) == jerry_get_number_value (jerry_get_property_by_index (typedarray, 2)));
+    TEST_ASSERT (!jerry_delete_property_by_index (typedarray, 1));
```
Comment (galpeter) [inline tests/unit-core/test-typedarray.c]: We should also test a case where we are removing an element on an unset index (for example: the 5th index was not set, but we still try to delete it)

File: tests/unit-core/test-typedarray.c [hunk 5]
```diff
@@ -303,6 +303,44 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+    jerry_value_t length_arg = jerry_create_number (test_entries[i].element_count);
+
+    jerry_value_t typedarray = jerry_construct_object (prop_value, &length_arg, 1);
+
+    jerry_release_value (prop_name);
+    jerry_release_value (prop_value);
+    jerry_release_value (length_arg);
+
+    jerry_set_property_by_index (typedarray, 0, ja);
+    jerry_set_property_by_index (typedarray, 1, jb);
```
Comment (galpeter) [inline tests/unit-core/test-typedarray.c]: We also need to test the case when the index is out of bounds, which should return what?

File: tests/unit-core/test-typedarray.c [hunk 6]
```diff
@@ -303,6 +303,44 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
 
+/**
+ * Test get/set/delete property by index.
+ */
+static void test_property_by_index (test_entry_t test_entries[])
+{
+
+  jerry_value_t ja = jerry_create_number (3);
+  jerry_value_t jb = jerry_create_number (6);
+  jerry_value_t jc = jerry_create_number (50);
```
Comment (galpeter) [inline tests/unit-core/test-typedarray.c]: We should improve this to test more values not just simple integers. We need negative numbers, zero, inf/-inf, double values. Having multiple values we can create an array of test values (similar to the test_entries array) thus we would not need multiple variables.

**Comment by imiklos** [comment]
@galpeter Thank you for the reviews. I updated the PR.

**Comment by imiklos** [comment]
@galpeter Thanks for the reviews. I updated the PR.

File: tests/unit-core/test-typedarray.c [hunk 7]
```diff
@@ -303,6 +303,102 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+    jerry_typedarray_type_t type = jerry_get_typedarray_type (typedarray);
+
+    switch (type)
+    {
+      case JERRY_TYPEDARRAY_INT8:
+      case JERRY_TYPEDARRAY_INT16:
+      case JERRY_TYPEDARRAY_INT32:
+      for (uint32_t j = 0; j < test_numbers_length; j+=2)
+      {
+        test_numbers[j] = jerry_create_number (0 - jerry_get_number_value (test_numbers[j]));
```
Comment (zherczeg) [inline tests/unit-core/test-typedarray.c]: What happens with the original number? I know you don't need to free them, but still you should according the documentation.

Btw, you could have a simple const number list, not an ecma_value_t list. This way you can avoid the release problem.

File: tests/unit-core/test-typedarray.c [hunk 8]
```diff
@@ -303,6 +303,102 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+      TEST_ASSERT (!jerry_delete_property_by_index (typedarray, j));
+      jerry_value_t set_result = jerry_set_property_by_index (typedarray, j, test_numbers[j]);
+      jerry_value_t get_result = jerry_get_property_by_index (typedarray, j);
+
+      TEST_ASSERT (jerry_value_is_boolean (set_result));
+      TEST_ASSERT (jerry_get_boolean_value (set_result));
+      TEST_ASSERT (!jerry_delete_property_by_index (typedarray, j));
+      if (type == JERRY_TYPEDARRAY_FLOAT32 || type == JERRY_TYPEDARRAY_FLOAT64)
+      {
+        double epsilon = pow (10,-7);
```
Comment (zherczeg) [inline tests/unit-core/test-typedarray.c]: Missing space before -7.

**Comment by imiklos** [comment]
@galpeter @zherczeg I updated the PR.

File: tests/unit-core/test-typedarray.c [hunk 9]
```diff
@@ -303,6 +303,107 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+      {
+        test_numbers[j] = jerry_create_number (test_double_numbers[j]);
+      }
+      break;
+      default:
+      for (uint8_t j = 0; j < test_numbers_length; j++)
+      {
+        test_numbers[j] = jerry_create_number (test_uint_numbers[j]);
+      }
+      break;
```
Comment (zherczeg) [inline tests/unit-core/test-typedarray.c]: Question: can we do this in one step? Create a single jerry value and set/get it?

**Comment by imiklos** [comment]
@zherczeg I updated the PR, made changes to set and get values in one step.

File: tests/unit-core/test-typedarray.c [hunk 10]
```diff
@@ -303,6 +303,115 @@ test_typedarray_complex_creation (test_entry_t test_entries[], /**< test cases *
...
+
+    jerry_value_t set_result;
+    jerry_value_t get_result;
+
+    switch (type)
+    {
+      case JERRY_TYPEDARRAY_INT8:
+      case JERRY_TYPEDARRAY_INT16:
+      case JERRY_TYPEDARRAY_INT32:
+      for (uint8_t j = 0; j < test_numbers_length; j++)
```
Comment (zherczeg) [inline tests/unit-core/test-typedarray.c]: I think the style excepts two spaces indentation after `case`

**Comment by imiklos** [comment]
@zherczeg I updated the PR, fixed the indentation after case.

**Pull request merged**